File: java/src/main/java/com/taobao/top/link/endpoint/ClientChannelSharedSelector.java
package com.taobao.top.link.endpoint;

import java.net.URI;
import java.util.Hashtable;

import com.taobao.top.link.channel.ChannelException;
import com.taobao.top.link.channel.ClientChannel;
import com.taobao.top.link.channel.ClientChannelSelector;
import com.taobao.top.link.channel.websocket.WebSocketClient;

public class ClientChannelSharedSelector implements ClientChannelSelector {
	private final static int CONNECT_TIMEOUT = 5000;
	private Hashtable<String, ClientChannel> channels;
	protected Object lockObject;

	public ClientChannelSharedSelector() {
		this.channels = new Hashtable<String, ClientChannel>();
		this.lockObject = new Object();
	}

	@Override
	public ClientChannel getChannel(URI uri) throws ChannelException {
		if (!uri.getScheme().equalsIgnoreCase("ws")) {
			return null;
		}
		final String url = uri.toString();
		if (channels.get(url) == null ||
				!channels.get(url).isConnected()) {
			synchronized (this.lockObject) {
				if (channels.get(url) == null ||
						!channels.get(url).isConnected()) {
					channels.put(url,
							WebSocketClient.connect(uri, CONNECT_TIMEOUT));
				}
			}
		}
		return channels.get(url);
	}

	@Override
	public void returnChannel(ClientChannel channel) {
		// shared channel
	}
}

File: csharp/src/Taobao.Top.Link/Channel/ClientChannelSharedSelector.cs
using System;
using System.Collections.Generic;
using System.Text;
using Taobao.Top.Link.Channel.WebSocket;

namespace Taobao.Top.Link.Channel
{
    /// <summary>a channel pool that same uri sharing same channel
    /// </summary>
    public class ClientChannelSharedSelector : IClientChannelSelector
    {
        private Object _lockObject;
        private IDictionary<string, IClientChannel> _channels;

        public ClientChannelSharedSelector()
        {
            this._lockObject = new object();
            this._channels = new Dictionary<string, IClientChannel>();
        }

        public IClientChannel GetChannel(Uri uri)
        {
            if (!uri.Scheme.Equals("ws", StringComparison.InvariantCultureIgnoreCase))
                return null;

            var url = uri.ToString();

            if (!this.HaveChannel(url))
                lock (this._lockObject)
                    if (!this.HaveChannel(url))
                        this._channels.Add(url, this.WrapChannel(this.Connect(uri, 5000)));

            return _channels[url];
        }
        public void ReturnChannel(IClientChannel channel) { }

        protected virtual IClientChannel Connect(Uri uri, int timeout)
        {
            return WebSocketClient.Connect(uri, timeout);
        }
        private IClientChannel WrapChannel(IClientChannel channel)
        {
            return channel;
        }
        private bool HaveChannel(string url)
        {
            return this._channels.ContainsKey(url) && this._channels[url].IsConnected;
        }
    }
}